crates/circuit/src/dag_circuit.rs

```diff
@@ -0,0 +1,6474 @@
+                .unwrap()
+                .map(|v| NodeIndex::new(v.extract().unwrap()));
+            let _del_result = bound_dict.del_item(key);
+            res
+        })
+    }
+    pub fn values<'py>(&self, py: Python<'py>) -> impl Iterator<Item = NodeIndex> + 'py {
+        let values = self.dict.bind(py).values();
+        values.iter().map(|x| NodeIndex::new(x.extract().unwrap()))
+    }
+}
+
+/// Quantum circuit as a directed acyclic graph.
+///
+/// There are 3 types of nodes in the graph: inputs, outputs, and operations.
+/// The nodes are connected by directed edges that correspond to qubits and
+/// bits.
+#[pyclass(module = "qiskit._accelerate.circuit")]
+#[derive(Clone, Debug)]
+pub struct DAGCircuit {
+    /// Circuit name.  Generally, this corresponds to the name
+    /// of the QuantumCircuit from which the DAG was generated.
+    #[pyo3(get, set)]
+    name: Option<PyObject>,
+    /// Circuit metadata
+    #[pyo3(get, set)]
+    metadata: Option<PyObject>,
+
+    calibrations: HashMap<String, Py<PyDict>>,
+
+    pub(crate) dag: StableDiGraph<NodeType, Wire>,
+
+    #[pyo3(get)]
+    qregs: Py<PyDict>,
+    #[pyo3(get)]
+    cregs: Py<PyDict>,
+
+    /// The cache used to intern instruction qargs.
+    qargs_cache: IndexedInterner<Vec<Qubit>>,
+    /// The cache used to intern instruction cargs.
+    cargs_cache: IndexedInterner<Vec<Clbit>>,
+    /// Qubits registered in the circuit.
+    pub(crate) qubits: BitData<Qubit>,
+    /// Clbits registered in the circuit.
+    pub(crate) clbits: BitData<Clbit>,
+    /// Global phase.
+    global_phase: Param,
+    /// Duration.
+    #[pyo3(get, set)]
+    duration: Option<PyObject>,
+    /// Unit of duration.
+    #[pyo3(get, set)]
+    unit: String,
+
+    // Note: these are tracked separately from `qubits` and `clbits`
+    // because it's not yet clear if the Rust concept of a native Qubit
+    // and Clbit should correspond directly to the numerical Python
+    // index that users see in the Python API.
+    /// The index locations of bits, and their positions within
+    /// registers.
+    qubit_locations: Py<PyDict>,
+    clbit_locations: Py<PyDict>,
+
+    /// Map from qubit to input nodes of the graph.
+    qubit_input_map: IndexMap<Qubit, NodeIndex, RandomState>,
+    /// Map from qubit to output nodes of the graph.
+    qubit_output_map: IndexMap<Qubit, NodeIndex, RandomState>,
+
+    /// Map from clbit to input nodes of the graph.
+    clbit_input_map: IndexMap<Clbit, NodeIndex, RandomState>,
+    /// Map from clbit to output nodes of the graph.
+    clbit_output_map: IndexMap<Clbit, NodeIndex, RandomState>,
+
+    // TODO: use IndexMap<Wire, NodeIndex> once Var is ported to Rust
+    /// Map from var to input nodes of the graph.
+    var_input_map: _VarIndexMap,
+    /// Map from var to output nodes of the graph.
+    var_output_map: _VarIndexMap,
+
+    /// Operation kind to count
+    op_names: IndexMap<String, usize, RandomState>,
```

Comment:
We have places that rely on the ordering indirectly at least. You can see the commit where I had to do this  here eccef5dbea6675efab9d2cc7c96f5c0d59c40ddc (from before the squash). The issue was there were places where the iteration order of `count_ops()` was influencing the substitution chosen and it was causing non-deterministic test results based on the hash seeding when I used a `HashMap`. Python side this was previously using `OrderedDict` instead of a `dict` too.  This was one of the last fixes needed for the tests.

Personally, I'm not super worried that the order isn't 100% the same exact behavior as the python implementation. I don't think we really offered the exact ordering across release boundaries as an api guarantee, as long as the iteration order is consistent and deterministic within a release.

Reply:
Ah ok, if it's reproducibility that's the important part rather than it being specifically insertion ordering, I'm fine with it.